### PR TITLE
[SYCL][NATIVECPU] Support multiple SYCL targets in the same compiler invocation

### PR DESCRIPTION
**Pull request by PietroGhg**
This PR adds support to multiple SYCL targets alongside `native_cpu` in the same compiler invocation (e.g. `clang++ -fsycl -fsycl-targets=native_cpu,spir64 input.cpp`). In order to implement this we had to make changes to multiple components, here is a quick overview:
* Driver: changes in the Driver allow to correctly parse all the targets passed to  `-fsycl-targets` (before we were just looking for `native_cpu`, ignoring the others). The Driver now also calls `sycl-post-link` and `clang-offload-wrapper`, performing a compilation flow more similar to the one used for other targets.
* Sema: since the kernel name needs to be the same for all the SYCL targets, the change to the kernel name in Sema has been removed, and replaced with an LLVM Pass that gets run when lowering the device module (`llvm/lib/SYCLLowerIR/RenameKernelSYCLNativeCPU.cpp`). 
* Runtime: The definition for `_pi_program` in the Native CPU Plug-In now supports multiple kernels in one program, and the `__SYCL_PI_DEVICE_BINARY_TARGET_NATIVE_CPU` binary type has been added in order to identify kernels compiled for Native CPU.
* clang-offload-wrapper: for Native CPU, the offload-wrapper doesn't bundle the device code in the host module, but instead produces an array containing function declarations that are resolved by the linker, see `sycl/doc/design/SYCLNativeCPU.md` for more information.

### CODE REVIEW

**Review by elizabethandrews — APPROVED** [review]
FE changes LGTM

**Review by mdtoguchi — APPROVED** [review]
OK for driver

**Review by uwedolinsky — APPROVED** [review]

**Comment by uwedolinsky** [comment]
Hello @intel/llvm-gatekeepers We think this PR is ready to be merged.

**Comment by steffenlarsen** [comment]
The PR still needs approval from the following groups:

- @intel/dpcpp-esimd-reviewers 
- @intel/dpcpp-l0-pi-reviewers 
- @intel/dpcpp-tools-reviewers 
- @intel/llvm-reviewers-runtime 

Ping to all these groups.

**Comment by dm-vodopyanov** [comment]
> Hello @intel/llvm-gatekeepers We think this PR is ready to be merged.

@uwedolinsky Many code reviews are expected: 
image
Please ping @intel/llvm-gatekeepers only after obtaining all required approvals.

@intel/dpcpp-esimd-reviewers @intel/dpcpp-l0-pi-reviewers @intel/dpcpp-tools-reviewers @bso-intel @intel/llvm-reviewers-runtime, can you please review?

**Comment by uwedolinsky** [comment]
> > Hello @intel/llvm-gatekeepers We think this PR is ready to be merged.
> 
> @uwedolinsky Many code reviews are expected: image Please ping @intel/llvm-gatekeepers only after obtaining all required approvals.

Apologies! I mistook "At least 1 approving review is required to merge this pull request" at the top to mean this PR can be merged because it already had approvals.

**Review by sarnex — APPROVED** [review]
esimd lgtm

**Review by bso-intel — APPROVED** [review]

File: clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp
Comment (AlexeySachkov) [inline clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp:1039]: Does this way of integrating `native_cpu` target allows for a single call to `clang-offload-wrapper` to wrap all targets at the same time?

I haven't reviewed the driver changes, but based on this, I assume that we invoking `clang-offload-wrapper` separately for `native_cpu`, right?

I wonder, can this `if` be implemented based on a target for the currently processed image?

**Review by AlexeySachkov — APPROVED** [review]
A few comments, but overall LGTM

File: clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp
Comment (uwedolinsky) [reply clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp]: > Does this way of integrating native_cpu target allows for a single call to clang-offload-wrapper to wrap all targets at the same time?

This should be possible but we haven't looked into that yet. But we'll put that on the todo list and create another PR with dedicated clang-offload-wrapper tests if that's ok?

> I haven't reviewed the driver changes, but based on this, I assume that we invoking clang-offload-wrapper separately for native_cpu, right?

Yes, that's right. However, it seems that for a normal clang invocation requesting multiple targets (even without native_cpu) the clang-offload-wrapper is currently invoked for each target individually. Or is there another option to generate only one call to `clang-offload-wrapper`?  

> I wonder, can this if be implemented based on a target for the currently processed image?

Great find! The option is now removed and the check is based on the image as you suggested. Thank You!

Comment (AlexeySachkov) [reply clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp]: > Yes, that's right. However, it seems that for a normal clang invocation requesting multiple targets (even without native_cpu) the `clang-offload-wrapper` is currently invoked for each target individually. Or is there another option to generate only one call to `clang-offload-wrapper`?

I honestly don't remember that implementation detail of the toolchain, better ask @mdtoguchi here. If we already launch `clang-offload-wrapper` individually for each target, then I don't think that we should do something special here for `native_cpu`.

My main point was to ensure that all targets are handled in a similar manner, that's all.

> But we'll put that on the todo list and create another PR with dedicated `clang-offload-wrapper` tests if that's ok?

As I've said, it is not even needed if we don't do  that for other targets already. I mean, for the purpose of optimizing the toolchain we may consider doing so, but I believe that it will be done automatically once we switch to `clang-linker-wrapper`. So, at the moment it is more of: we need to keep that in mind when further working on `native_cpu` target

Comment (mdtoguchi) [reply clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp]: The wrapper is called for each target individually, so this is fine.  There is a way to do a singular call for multiple targets, but the current toolchain is not setup to take advantage of this at the moment.

Comment (uwedolinsky) [reply clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp]: Thank you very much @AlexeySachkov for raising this and @mdtoguchi for the info, we'll keep an eye on this. If `native_cpu` causes any issues with `clang-offload-wrapper` or `clang-linker-wrapper`  please let us know. Thank you!

**Comment by uwedolinsky** [comment]
Hello @intel/dpcpp-l0-pi-reviewers, please could you review this PR? Thank you!

File: sycl/plugins/unified_runtime/pi2ur.hpp
Comment (smaslov-intel) [inline sycl/plugins/unified_runtime/pi2ur.hpp:1342]: Did you request this addition from the UR?
Tag @kbenzie

File: sycl/plugins/unified_runtime/ur/adapters/native_cpu/program.cpp
Comment (smaslov-intel) [inline sycl/plugins/unified_runtime/ur/adapters/native_cpu/program.cpp:31]: shouldn't this definition come from some header file?

File: sycl/plugins/unified_runtime/pi2ur.hpp
Comment (uwedolinsky) [reply sycl/plugins/unified_runtime/pi2ur.hpp:1342]: I've created a UR PR for this: https://github.com/oneapi-src/unified-runtime/pull/753

File: sycl/plugins/unified_runtime/ur/adapters/native_cpu/program.cpp
Comment (PietroGhg) [reply sycl/plugins/unified_runtime/ur/adapters/native_cpu/program.cpp]: Yeah thanks for pointing it out, I moved the definition to `program.hpp`

File: sycl/plugins/unified_runtime/ur/adapters/native_cpu/program.hpp
Comment (smaslov-intel) [inline sycl/plugins/unified_runtime/ur/adapters/native_cpu/program.hpp:35]: This certainly deserved some comment about what it is an how it is used. But isn't it defined somewhere else already? If we can't use that master definition directly here, then this should at least be a a big note that we need to keep in sync.

Comment (uwedolinsky) [reply sycl/plugins/unified_runtime/ur/adapters/native_cpu/program.hpp:35]: This struct is only used for kernel lookup inside `program.cpp` which is why its definition was "hidden" inside that file before this change. Only the OffloadWrapper creates instances of this struct, but it does that directly in LLVM IR which is why there is no C++ definition of that struct anywhere else.

Comment (smaslov-intel) [reply sycl/plugins/unified_runtime/ur/adapters/native_cpu/program.hpp:35]: > Only the OffloadWrapper creates instances of this struct, but it does that directly in LLVM IR

Could you point where this is done so I could better understand why it cannot be shared?
Also, please do add a source code comment about this.

Comment (PietroGhg) [reply sycl/plugins/unified_runtime/ur/adapters/native_cpu/program.hpp:35]: The struct is defined in the offload wrapper [here](https://github.com/PietroGhg/llvm/blob/de4ca64f479ea627521c20cb6d1571a0cb2fc874/clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp#L631), it is a pattern followed by the other structs used by PI e.g. the [offload_entry](https://github.com/PietroGhg/llvm/blob/de4ca64f479ea627521c20cb6d1571a0cb2fc874/clang/tools/clang-offload-wrapper/ClangOffloadWrapper.cpp#L390) struct, the definition can't be directly shared in a straight forward way. I had the comment in the offload wrapper but not in the Plug-in code, I'll add the comment here since as you say it is much needed, thank you

Comment (smaslov-intel) [reply sycl/plugins/unified_runtime/ur/adapters/native_cpu/program.hpp:35]: Then maybe move it to pi.h around here: https://github.com/intel/llvm/blob/a8cc71eb55171cef613d8e1077870fb0ea441b14/sycl/include/sycl/detail/pi.h#L953
?

Comment (PietroGhg) [reply sycl/plugins/unified_runtime/ur/adapters/native_cpu/program.hpp:35]: While I see the benefit of having all the structs needed by the offload-wrapper defined in the same file, maybe it would be better to keep the `nativecpu_entry` `struct defined in the Native CPU adapter, without exposing it to all the PI plug-ins. E.g. if another contributor makes a change to that struct we (Native CPU) maintainers would not get notified if we move it to `pi.h`, I think

Comment (smaslov-intel) [reply sycl/plugins/unified_runtime/ur/adapters/native_cpu/program.hpp:35]: I am OK to keep it where it is, just add a comment

Comment (PietroGhg) [reply sycl/plugins/unified_runtime/ur/adapters/native_cpu/program.hpp:39]: I added the comment, thank you

**Comment by PietroGhg** [comment]
@intel/llvm-gatekeepers this looks ready to be merged thank you :)

**Pull request merged**
